c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c

```diff
@@ -339,3 +340,359 @@ DWORD request_sys_process_memory_unlock(Remote *remote, Packet *packet)
+typedef const UNICODE_STRING* PCUNICODE_STRING;
+
+// https://learn.microsoft.com/en-us/windows/win32/api/ntdef/ns-ntdef-_object_attributes
+typedef struct _OBJECT_ATTRIBUTES {
+	ULONG           Length;
+	HANDLE          RootDirectory;
+	PUNICODE_STRING ObjectName;
+	ULONG           Attributes;
+	PVOID           SecurityDescriptor;
+	PVOID           SecurityQualityOfService;
+} OBJECT_ATTRIBUTES, * POBJECT_ATTRIBUTES;
+
+// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tsts/a11e7129-685b-4535-8d37-21d4596ac057
+typedef struct _CLIENT_ID {
+	HANDLE UniqueProcess;
+	HANDLE UniqueThread;
+} CLIENT_ID, * PCLIENT_ID;
+
+// http://undocumented.ntinternals.net/index.html?page=UserMode%2FUndocumented%20Functions%2FNT%20Objects%2FProcess%2FNtOpenProcess.html
+// https://ntdoc.m417z.com/ntopenprocess
+typedef NTSTATUS(NTAPI* NTOPENPROCESS) (PHANDLE ProcessHandle, ACCESS_MASK AccessMask, POBJECT_ATTRIBUTES ObjectAttributes, PCLIENT_ID ClientId);
+
+//typedef struct _PEB_LDR_DATA //, 7 elements, 0x28 bytes
+//{
+//	DWORD dwLength;
+//	DWORD dwInitialized;
+//	LPVOID lpSsHandle;
+//	LIST_ENTRY InLoadOrderModuleList;
+//	LIST_ENTRY InMemoryOrderModuleList;
+//	LIST_ENTRY InInitializationOrderModuleList;
+//	LPVOID lpEntryInProgress;
+//} PEB_LDR_DATA, * PPEB_LDR_DATA;
+
+typedef struct _RTL_USER_PROCESS_PARAMETERS {
+	BYTE Reserved1[16];
+	PVOID Reserved2[10];
+	UNICODE_STRING ImagePathName;
+	UNICODE_STRING CommandLine;
+} RTL_USER_PROCESS_PARAMETERS, * PRTL_USER_PROCESS_PARAMETERS;
+
+typedef
+VOID
+(NTAPI* PPS_POST_PROCESS_INIT_ROUTINE) (
+	VOID
+	);
+
+typedef struct _PEB {
+	BYTE Reserved1[2];
+	BYTE BeingDebugged;
+	BYTE Reserved2[1];
+	PVOID Reserved3[2];
+	PPEB_LDR_DATA Ldr;
+	PRTL_USER_PROCESS_PARAMETERS ProcessParameters;
+	BYTE Reserved4[104];
+	PVOID Reserved5[52];
+	PPS_POST_PROCESS_INIT_ROUTINE PostProcessInitRoutine;
+	BYTE Reserved6[128];
+	PVOID Reserved7[1];
+	ULONG SessionId;
+} PEB, * PPEB;
+
+typedef struct _PROCESS_BASIC_INFORMATION {
+	PVOID Reserved1;
+	PPEB PebBaseAddress;
+	PVOID Reserved2[2];
+	ULONG_PTR UniqueProcessId;
+	PVOID Reserved3;
+} PROCESS_BASIC_INFORMATION;
+typedef PROCESS_BASIC_INFORMATION* PPROCESS_BASIC_INFORMATION;
+
+typedef enum _PROCESSINFOCLASS {
+	ProcessBasicInformation = 0,
+	ProcessWow64Information = 26
+} PROCESSINFOCLASS;
+
+BOOL can_read_memory(DWORD memory_protect)
+{
+	const int page_execute_read = 0x20;
+	const int page_execute_readwrite = 0x40;
+	const int page_readonly = 0x02;
+	const int page_readwrite = 0x04;
+
+	return memory_protect == page_execute_read ||
+		memory_protect == page_execute_readwrite ||
+		memory_protect == page_readonly ||
+		memory_protect == page_readwrite;
+}
+
+// In order to be able to regex null-butes, we need to store the length explicitly, so that null-bytes aren't being treated as the end of a string.
+struct regex_needle
+{
+	char* raw_needle_buffer;
+	size_t length;
+	regex_t* compiled_needle;
+};
+
+#define NEEDLES_MAX (size_t)5

```

Comment:
Is this limit enforced in Metasploit/mettle somewhere? It seems a bit arbitrary 👀

Reply:
It's being used in Mettle: https://github.com/rapid7/mettle/blob/a2f3c756ea84f07b80d64ff815cfec2b2e6d27e3/mettle/src/stdapi/sys/memory.c#L15 👍